jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/ServletChannel.java

```diff
@@ -58,31 +58,42 @@
 import static org.eclipse.jetty.util.thread.Invocable.InvocationType.NON_BLOCKING;
 
 /**
- * TODO describe what this class does
+ * The ServletChannel contains the state and behaviors associated with the Servlet API
+ * lifecycle for a single request/response cycle. Specifically it uses
+ * {@link ServletRequestState} to coordinate the states of dispatch state, input and
+ * output according to the servlet specification.  The combined state so obtained
+ * is reflected in the behaviour of the contained {@link HttpInput} implementation of
+ * {@link jakarta.servlet.ServletInputStream}.
+ *
+ * @see ServletRequestState
+ * @see HttpInput
  */
-public class ServletChannel implements Runnable
+public class ServletChannel
 {
     private static final Logger LOG = LoggerFactory.getLogger(ServletChannel.class);
 
     private final ServletRequestState _state;
+    private final ServletContextHandler.Context _context;
+    private final ServletContextHandler.ServletContextApi _servletContextApi;
     private final AtomicLong _requests = new AtomicLong();
     private final Connector _connector;
     private final Executor _executor;
     private final HttpConfiguration _configuration;
     private final EndPoint _endPoint;
     private final HttpInput _httpInput;
     private final Listener _combinedListener;
-    private ServletContextHandler.ServletContextApi _servletContextApi;
-    private ServletContextRequest _servletContextRequest;
-    private boolean _expects100Continue;
-    private long _oldIdleTimeout;
-    private Callback _callback;
+    private volatile ServletContextRequest _servletContextRequest;
+    private volatile boolean _expects100Continue;
+    private volatile long _oldIdleTimeout;
+    private volatile Callback _callback;
     // Bytes written after interception (e.g. after compression).
-    private long _written;
+    private volatile long _written;
 
     public ServletChannel(ServletContextHandler servletContextHandler, Request request)
     {
         _state = new ServletRequestState(this);
+        _context = servletContextHandler.getContext();
+        _servletContextApi = _context.getServletContext();
```

Comment:
Couldn't this be recycled and go through a different `ServletContextHandler` which would have a different `ServletContextApi`? 

If so then this would need to be done in `associate()`.

Reply:
Wouldn't it be better to have a `ServletChannel` which could be re-used for any context instead of always remaking it if it goes to a different context.

Reply:
Yeah maybe, but as @lorban would say, that is optimizing the slow path.  Cross context requests are rare and many deployments only have 1 context per server.
Having these big objects with lots of final fields increases the benefit of re-using them, as all those final fields don't need to be re-enabled.

So I'm happy with it as is for now.   Perhaps if we see cross context slowness as a problem in future we invest in a reusable instance per context?

Reply:
Ok, at least its better than before so I'll approve.

I can't see that there would be much extra cost to making it fully re-usable though.